[WIP] Send more data frsky telemetry

thedevleon opened the pull request:
I've been working on this the last couple of days and managed to implement the sending of all kinds of data that might be useful to be displayed on a sPort and openTX 2.2 compatible radio either by itself, or via scripts (like my https://github.com/thedevleon/PixPilot, which is also still heavily WIP)

New data that will be sent:
- Acceleration in X,Y and Z
- Arming State
- Attitude Roll, Pitch and Yaw
- Mission current & reached sequence as well as it's status
- Mavlink Messages
- Average Cells Voltage
- GPS Time and Date

This includes a couple of fixes as well I made as I got more familiar with how opentx handles telemetry.
- all data values should now be sent correctly (like acceleration, gps date and time, and more)
- corrected frequencies for sending data
- instead of offsetting the nav_state by 128, I offset all the sensor id's for PX4 by 128. That way 0-128 can be used by ardupilot and we use 128-255, and we have the full range of the uint32_t because there is no further offsets.

This is still WIP - I'm looking for feedback if there is anything else that should be sent.

The DIY values all require OpenTX 2.2 of course.
For reference on how to use these, please take a look at these links:
- https://github.com/PX4/Firmware/pull/4349
- https://github.com/opentx/opentx/pull/3326
- https://github.com/opentx/opentx/pull/3426
#### Progress:
- [x] add and verify id's for wanted data
- [x] functions for wanted data
- [x] verify receiving of wanted data
- [x] implement sending of mavlink messages
- [ ] verify correct format of values and display in opentx

LorenzMeier commented:
Can you run "top" on the shell and validate that your FrSky app is not consuming more memory than it has?

LorenzMeier commented:
Otherwise looking neat!

thedevleon commented:
Output from top:
`222 frsky_telemetry 2844 2.646 756/ 1096 200 (200) w:sem`

Full top:
![top](http://i.imgur.com/Z7zyL9w.png)

LorenzMeier commented on src/drivers/frsky_telemetry/frsky_data.c:
```diff
@@ -218,11 +218,11 @@ void frsky_send_frame1(int uart)
+			roundf(sensor_combined->accelerometer_m_s2[0] * 1000.0f * (1 / 9.81f)));
 	frsky_send_data(uart, FRSKY_ID_ACCEL_Y,
-			roundf(sensor_combined->accelerometer_m_s2[1] * 1000.0f));
+			roundf(sensor_combined->accelerometer_m_s2[1] * 1000.0f * 1000.0f * (1 / 9.81f)));
```
Why is the X, Y and Z scaling different?

thedevleon replied:
According to the OpenTX source [here](https://github.com/opentx/opentx/blob/dbd8abbfe8343b5d7f542304d47e232140307b95/radio/src/telemetry/frsky_d_arm.cpp#L122) and [here](https://github.com/opentx/opentx/blob/65d7fd0a290b48af2a9a8d9f9be076905a2c8e4a/radio/src/telemetry/frsky_sport.cpp#L74), these values are expected to be in g.
Since accelerometer_m_s2 is in m/s^2 I'm multiplying by 1/9.81m/s^2 to get the value in g.

kd0aij commented:
there's a lightweight C ringbuffer in src/modules/sdlog2/logbuffer.c/h
It's well tested.

kd0aij commented:
I don't think GPLv2 is compatible with the PX4 license

LorenzMeier commented:
No, we will not be able to accept changes including GPLv2 code because it would mean that the whole codebase would have to become GPLv2.

thedevleon commented:
Oh okay no worries, I'll find a different solution.

thedevleon commented:
I've implemented by own fifo with the great help of @dustingooding.

thedevleon commented:
Everything works now :) Benchmarks so far look good!

thedevleon commented:
A small lua script for testing the sending of the mavlink messages can be found here:
https://gist.github.com/thedevleon/b2d5996cf6c48e8d3fa659d0a865ff40

LorenzMeier commented:
Nice! Could you do a video showing what it transmits now and how it looks? I'm sure the community would love that.

thedevleon commented:
@LorenzMeier Sure! I'll be back home soon (currently on travels), and will post a video then. I'll also add some documentation to the gitbook about this :)

thedevleon commented:
https://github.com/opentx/opentx/pull/3426 has just been merged, will have to take a closer look and take those changes into consideration again. Seems like the structure of the input and output lua fifo have been changed.

thedevleon commented:
Everything seems to be in order now. I will be testing if all the telemetry values are being displayed correctly in opentx in the next fews days, since I just finished my quad. 
I also have made some big changes and progress with [PixPilot](https://github.com/thedevleon/PixPilot)!

LorenzMeier commented on src/drivers/frsky_telemetry/sPort_data.c:
```diff
@@ -64,13 +67,24 @@ static int global_position_sub = -1;
+//FIFO for mavlink messages
+#define FIFO_ELEMENTS 636 //12*52, so up to 12 messages, worst case.
+#define FIFO_SIZE (FIFO_ELEMENTS + 1)
+uint8_t fifo[FIFO_SIZE];
```
This will eat the RAM globally, even if its not used. And its a pretty big allocation. This should be done inside the driver, not globally, and 600 bytes seems way too much. Try to cut it down to 3-4 messages.

thedevleon commented:
@LorenzMeier I adjusted the fifo. Does it look good now?

LorenzMeier commented:
Pretty nice! If I don't have the LUA script installed yet, what will I get? Consistent data, but some (DIY) fields missing?

thedevleon commented:
@LorenzMeier you will get VFAS (Battery Voltage), CELLS, CUR (Current), ALT (Altitude), VSPD (Vertical Speed), FUEL, ACCX, ACCY, ACCZ, GPS (Lon, Lat, Alt, Spd, Crs, Date and Time) all via the normal SPort Data ids - so those can be viewed directly in the radio and can be assigned to telemetry screens without any scripts.
gps_fix, nav_state, arming_state, roll, pitch, yaw, current mission sequence, reached mission sequence, current mission status flags and mavlink messages are all sent via DIY values - and thus need scripts to interpret them (but they are all almost already implemented and parsed by the PixPilot PX4 adapter).

kd0aij commented:
@thedevleon There's a report here: http://discuss.px4.io/t/broken-frsky-telemetry-on-sport-x8r/462
that something is broken.

thedevleon commented:
@kd0aij thanks! I did reply there. I really need to confirm and finish reverse engineering the way opentx decodes the data we want to sent (especially GPS time/date/lot/lan/alt and cells, etc..). Help with that and testing will always be appreciated :)
The openTx class that parses all the telemetry values can be found here: https://github.com/opentx/opentx/blob/master/radio/src/telemetry/telemetry.cpp

thedevleon commented:
I added comments to the functions that now need to be verified as sending the data correctly.

Those are the following:
* sPort_send_CELLS
* sPort_send_ACCX
* sPort_send_ACCY
* sPort_send_ACCZ
* sPort_send_GPS_LON
* sPort_send_GPS_LAT
* sPort_send_GPS_DATE
* sPort_send_GPS_TIME

Everything else should be sending correctly.
As soon as I have verified and corrected all the functions above, this should ready to merge.

LorenzMeier commented:
@thedevleon Are you back active? I think we need a complete tutorial for the transmitter setup in our user guide on http://dev.px4.io. If you send me an email with your preferred email address to lorenz@px4.io I can set up an account in Wordpress for you.

thedevleon commented:
@LorenzMeier Just had my last exam today, will be able to further work on this in the next few days :) Sent you an email as well.

LorenzMeier commented:
Awesome, looking forward to it!

LorenzMeier commented:
@thedevleon @kd0aij What is the status here? I would either like to get this in and well documented or to revert to OpenTX standard fields so that no RX-side scripting is required.

LorenzMeier commented:
@thedevleon I'm closing this as abandoned. I still would love to merge this, but it needs to be finished and documented.
